engines/qemu/metaservice/shell.go

```diff
@@ -0,0 +1,323 @@
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
+)
+
+const (
+	// ShellHandshakeTimeout is the maximum allowed time for websocket handshake
+	ShellHandshakeTimeout = 3 * time.Second
+	// ShellPingInterval is the time between sending pings
+	ShellPingInterval = 5 * time.Second
+	// ShellWriteTimeout is the maximum time between successful writes
+	ShellWriteTimeout = ShellPingInterval * 2
+	// ShellPongTimeout is the maximum time between successful reads
+	ShellPongTimeout = ShellPingInterval * 3
+	// ShellBlockSize is the maximum number of bytes to send in a single block
+	ShellBlockSize = 16 * 1024
+	// ShellMaxMessageSize is the maximum message size we will read
+	ShellMaxMessageSize = ShellBlockSize + 4*1024
+	// ShellMaxPendingBytes is the maximum number of bytes allowed in-flight
+	ShellMaxPendingBytes = 4 * ShellBlockSize
+)
+
+type shell struct {
+	ws           *websocket.Conn
+	mWrite       sync.Mutex
+	stdin        io.WriteCloser
+	stdout       io.ReadCloser
+	stderr       io.ReadCloser
+	stdinReader  *ioext.PipeReader
+	stdoutWriter io.WriteCloser
+	stderrWriter io.WriteCloser
+	resolve      atomics.Once // Must wrap access to success/err
+	success      bool
+	err          error
+}
+
+// newShell takes a websocket and creates a shell object implementing the
+// engines.Shell interface.
+func newShell(ws *websocket.Conn) *shell {
+	stdinReader, stdin := ioext.BlockedPipe()
+	tellOut := make(chan int, 10)
+	tellErr := make(chan int, 10)
+	stdout, stdoutWriter := ioext.AsyncPipe(ShellMaxPendingBytes, tellOut)
+	stderr, stderrWriter := ioext.AsyncPipe(ShellMaxPendingBytes, tellErr)
+	stdinReader.Unblock(ShellMaxPendingBytes)
+
+	s := &shell{
+		ws:           ws,
+		stdin:        stdin,
+		stdout:       stdout,
+		stderr:       stderr,
+		stdinReader:  stdinReader,
+		stdoutWriter: stdoutWriter,
+		stderrWriter: stderrWriter,
+	}
+
+	ws.SetReadLimit(ShellMaxMessageSize)
+	ws.SetReadDeadline(time.Now().Add(ShellPongTimeout))
+	ws.SetPongHandler(s.pongHandler)
+
+	go s.writeMessages()
+	go s.readMessages()
+	go s.sendPings()
+	go s.sendAck(StreamStdout, tellOut)
+	go s.sendAck(StreamStderr, tellErr)
+
+	return s
+}
+
+func (s *shell) dispose() {
+	// Close websocket
+	s.ws.Close()
+
+	// Close all streams
+	s.stdinReader.Close()
+	s.stdoutWriter.Close()
+	s.stderrWriter.Close()
+}
+
+func (s *shell) send(message []byte) {
+	// Write message and ensure we reset the write deadline
+	s.mWrite.Lock()
+	s.ws.SetWriteDeadline(time.Now().Add(ShellWriteTimeout))
+	err := s.ws.WriteMessage(websocket.BinaryMessage, message)
+	s.mWrite.Unlock()
+
+	if err != nil {
+		s.resolve.Do(func() {
+			debug("Resolving internal error: Failed to send message, error: %s", err)
+			s.success = false
+			s.err = engines.ErrNonFatalInternalError
+			s.dispose()
+		})
+	}
+}
+
+func (s *shell) sendPings() {
+	for {
+		// Sleep for ping interval time
+		time.Sleep(ShellPingInterval)
```

Comment:
yeah, that is nice... except it leaks! You can't stop the ticker ever... See:
https://golang.org/pkg/time/#Tick